File: lib/src/ast/selector/id.dart
// Copyright 2016 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'dart:math' as math;

import 'package:meta/meta.dart';
import 'package:source_span/source_span.dart';

import '../../visitor/interface/selector.dart';
import '../selector.dart';

/// An ID selector.
///
/// This selects elements whose `id` attribute exactly matches the given name.
///
/// {@category AST}
@sealed
class IDSelector extends SimpleSelector {
  /// The ID name this selects for.
  final String name;

  int get specificity => math.pow(super.specificity, 2) as int;

  IDSelector(this.name, FileSpan span) : super(span);

  T accept<T>(SelectorVisitor<T> visitor) => visitor.visitIDSelector(this);

  /// @nodoc
  @internal
  IDSelector addSuffix(String suffix) => IDSelector(name + suffix, span);

  /// @nodoc
  @internal
  List<SimpleSelector>? unify(List<SimpleSelector> compound) {
    // A given compound selector may only contain one ID.
    if (compound.any((simple) => simple is IDSelector && simple != this)) {
      return null;
    }

    return super.unify(compound);
  }

  bool operator ==(Object other) => other is IDSelector && other.name == name;

  int get hashCode => name.hashCode;
}

File: pkg/sass-parser/lib/src/selector/id.ts
// Copyright 2025 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import * as postcss from 'postcss';

import {Interpolation, InterpolationProps} from '../interpolation';
import {LazySource} from '../lazy-source';
import type {AnyNode, NodeProps} from '../node';
import type {AnyStatement} from '../statement';
import * as sassInternal from '../sass-internal';
import * as utils from '../utils';
import {SimpleSelector} from './index';

/**
 * The initializer properties for {@link IDSelector}.
 *
 * @category Selector
 */
export interface IDSelectorProps extends NodeProps {
  id: Interpolation | InterpolationProps;
  raws?: IDSelectorRaws;
}

/**
 * Raws indicating how to precisely serialize an {@IDSelector}.
 *
 * @category Selector
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface -- No raws for an ID selector yet.
export interface IDSelectorRaws {}

/**
 * An ID selector.
 *
 * This selects elements whose `id` attribute contains an identifier with the
 * given name.
 *
 * @category Selector
 */
export class IDSelector extends SimpleSelector {
  readonly sassType = 'id' as const;
  declare raws: IDSelectorRaws;

  /** The ID name that this selects. */
  get id(): Interpolation {
    return this._id;
  }
  set id(id: Interpolation | InterpolationProps) {
    if (this._id) this._id.parent = undefined;
    const built =
      typeof id === 'object' && 'sassType' in id ? id : new Interpolation(id);
    built.parent = this;
    this._id = built;
  }
  private declare _id: Interpolation;

  constructor(defaults: IDSelectorProps);
  /** @hidden */
  constructor(_: undefined, inner: sassInternal.IDSelector);
  constructor(defaults?: object, inner?: sassInternal.IDSelector) {
    super(defaults);
    if (inner) {
      this.source = new LazySource(inner);
      this.id = new Interpolation(undefined, inner.name);
    }
  }

  clone(overrides?: Partial<IDSelectorProps>): this {
    return utils.cloneNode(this, overrides, ['raws', 'id']);
  }

  toJSON(): object;
  /** @hidden */
  toJSON(_: string, inputs: Map<postcss.Input, number>): object;
  toJSON(_?: string, inputs?: Map<postcss.Input, number>): object {
    return utils.toJSON(this, ['id'], inputs);
  }

  /** @hidden */
  toString(): string {
    return `#${this.id}`;
  }

  /** @hidden */
  get nonStatementChildren(): ReadonlyArray<Exclude<AnyNode, AnyStatement>> {
    return [this.id];
  }
}
